src/stream/src/executor/sink.rs

```diff
@@ -190,79 +189,103 @@ impl<F: LogStoreFactory> SinkExecutor<F> {
         }
     }
 
-    async fn execute_consume_log<R: LogReader>(
-        config: SinkConfig,
-        schema: Schema,
-        pk_indices: Vec<usize>,
-        connector_params: ConnectorParams,
-        sink_type: SinkType,
+    async fn execute_consume_log<S: Sink, R: LogReader>(
+        mut sink: S,
         mut log_reader: R,
         sink_commit_duration_metrics: Histogram,
     ) -> StreamExecutorResult<Message> {
-        let mut sink = build_sink(config, schema, pk_indices, connector_params, sink_type).await?;
+        log_reader.init().await?;
 
-        let mut epoch = log_reader.init().await?;
+        enum SinkState {
+            Uninitialized,
+            Writing { curr_epoch: u64 },
+            Checkpointed { prev_epoch: u64 },
+        }
 
-        // the flag is required because kafka transaction requires at least one
-        // message, so we should abort the transaction if the flag is true.
-        let mut empty_checkpoint_flag = true;
-        let mut in_transaction = false;
+        let mut state = SinkState::Uninitialized;
 
         loop {
-            let item: LogStoreReadItem = log_reader.next_item().await?;
+            let (epoch, item): (u64, LogStoreReadItem) = log_reader.next_item().await?;
             match item {
                 LogStoreReadItem::StreamChunk(chunk) => {
                     // NOTE: We start the txn here because a force-append-only sink might
                     // receive a data chunk full of DELETE messages and then drop all of them.
                     // At this point (instead of the point above when we receive the upstream
                     // data chunk), we make sure that we do have data to send out, and we can
                     // thus mark the txn as started.
```

Comment:
We can remove the comments here, since we're guaranteed to have data when we receive `LogStoreReadItem::StreamChunk` from the log store.

Reply:
Have removed the comments.